[BUG] Have a retry on 504 gateway timeout errors when polling on a resource state

### Is there an existing issue for this?

- [x] I have searched the existing issues

### What version of the Terraform provider are you using?

1.8.0

### What version of the Terraform CLI are you using?

1.9.3

### What type of issue are you facing

bug report

### Describe the bug

Depending on the capacity pressure combined with the network distance of the btp region it may happen the BTP CLI will timeout with 504 gateway timeout error causing the abort of a terraform apply run.

The BTP resource in question (most of often seen with SWZ subscriptions and kyma runtime envs) continues to be created asynchronously as nothing happened. 

However the btp provider return an error and the whole apply run fails.

### Expected Behavior

Add more resilience  (timeout, with a number of retries, and a circuit breaker) when it comes to resources state polling scenarios

### Steps To Reproduce

_No response_

### User's Role Collections

_No response_

### Add screenshots to help explain your problem

_No response_

### Additional context

_No response_

@prajin-op Thanks for the heads up. 
I think 504 errors must be mitigated by the provider itself.

However, from time to time there is this dreadful 500 error, due to probably a greater network latency (for instance running a tf script from Europe against a BTP landscape in Asia). 

Still, this has no adverse impact on the runtime creation which heads to its completion....on the BTP side.

Looks like runtime status polling with the BTP CLI APIs is not a very robust operation.

```
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [5m40s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [5m50s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m0s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m10s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m20s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m30s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m40s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m50s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m0s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m10s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m20s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m30s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m40s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m50s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [8m0s elapsed]
╷
│ Error: API Error Creating Resource Environment Instance (Subaccount)
│
│   with module.runtime_context.btp_subaccount_environment_instance.kyma[0],
│   on .terraform/modules/runtime_context/btp-context/runtime-context/modules/kyma-runtime/bootstrap-kymaruntime.tf line 258, in resource "btp_subaccount_environment_instance" "kyma":
│  258: resource "btp_subaccount_environment_instance" "kyma" {
│
│ received response with unexpected status [Status: 500; Correlation ID:
│ 97b83bdd-2a7c-8398-866b-ed3d857a46e8]
╵
Operation failed: failed running terraform apply (exit 1)
Error: Process completed with exit code 1.
```

All in all, the outcome is that the kyma runtime provisioning with tf btp provider is hugely unreliable.....regardless of the origin of the errors.

The BTP CLI polling API should not be trusted --> thus, a custom logic of retries, timeouts and circuit breakers should be implemented within the btp provider to address it.

In the meantime, all I can do is to revert to BTP CLI shell scripting and implement my safety net logic there.
Any other way around it ? Please advise.

--- a/internal/provider/resource_subaccount_entitlement.go
+++ b/internal/provider/resource_subaccount_entitlement.go
@@ -166,7 +166,7 @@ func (rs *subaccountEntitlementResource) Read(ctx context.Context, req resource.
 
 			entitlement, _, err := rs.cli.Accounts.Entitlement.GetAssignedBySubaccount(ctx, state.SubaccountId.ValueString(), state.ServiceName.ValueString(), state.PlanName.ValueString(), isParentGlobalAccount, parentId)
 
-			if isRetriableError(err) {
+			if tfutils.IsRetriableErrorForEntitlement(err) {
 				return nil, cis_entitlements.StateProcessing, nil
 			}
 
@@ -254,7 +254,7 @@ func (rs *subaccountEntitlementResource) createOrUpdate(ctx context.Context, req
 				return callResult, entitlementCallRetrySucceeded, nil
 			}
 
-			if isRetriableError(err) {
+			if tfutils.IsRetriableErrorForEntitlement(err) {
 				return callResult, entitlementCallRetryPending, nil
 			}
 
@@ -287,7 +287,7 @@ func (rs *subaccountEntitlementResource) createOrUpdate(ctx context.Context, req
 
 			entitlement, _, err := rs.cli.Accounts.Entitlement.GetAssignedBySubaccount(ctx, plan.SubaccountId.ValueString(), plan.ServiceName.ValueString(), plan.PlanName.ValueString(), isParentGlobalAccount, parentId)
 
-			if isRetriableError(err) {
+			if tfutils.IsRetriableErrorForEntitlement(err) {
 				return nil, cis_entitlements.StateProcessing, nil
 			}
 
@@ -361,7 +361,7 @@ func (rs *subaccountEntitlementResource) Delete(ctx context.Context, req resourc
 				return callResult, entitlementCallRetrySucceeded, nil
 			}
 
-			if isRetriableError(err) {
+			if tfutils.IsRetriableErrorForEntitlement(err) {
 				return callResult, entitlementCallRetryPending, nil
 			}
 
@@ -397,7 +397,7 @@ func (rs *subaccountEntitlementResource) Delete(ctx context.Context, req resourc
 				return entitlement, "DELETED", nil
 			}
 
-			if isRetriableError(err) {
+			if tfutils.IsRetriableErrorForEntitlement(err) {
 				return nil, cis_entitlements.StateProcessing, nil
 			}
 
@@ -455,24 +455,4 @@ func hasPlanQuota(state subaccountEntitlementType) bool {
 	}
 
 	return true
-}
-
-func isRetriableError(err error) bool {
-
-	if err == nil {
-		//If no error was raised a retry is not necessary
-		return false
-	}
-
-	if strings.Contains(err.Error(), "[Error: 30004/400]") {
-		// Error code for a locking scenario - API call must be retried
-		return true
-	} else if strings.Contains(err.Error(), "[Error: 11006/429]") {
-		// Error code when hitting a rate limit - API call must be retried
-		return true
-	} else {
-		// No retry possible as error code does not indicate a valid retry scenario
-		return false
-	}
-
 }

--- a/internal/provider/resource_subaccount_environment_instance.go
+++ b/internal/provider/resource_subaccount_environment_instance.go
@@ -270,6 +270,10 @@ func (rs *subaccountEnvironmentInstanceResource) Create(ctx context.Context, req
 		Refresh: func() (interface{}, string, error) {
 			subRes, _, err := rs.cli.Accounts.EnvironmentInstance.Get(ctx, plan.SubaccountId.ValueString(), cliRes.Id)
 
+			if tfutils.IsRetriableErrorForEnvInstance(err) {
+				return nil, provisioning.StateCreating, nil
+			}
+
 			if err != nil {
 				return subRes, "", err
 			}
@@ -326,6 +330,10 @@ func (rs *subaccountEnvironmentInstanceResource) Update(ctx context.Context, req
 		Refresh: func() (interface{}, string, error) {
 			subRes, _, err := rs.cli.Accounts.EnvironmentInstance.Get(ctx, plan.SubaccountId.ValueString(), plan.Id.ValueString())
 
+			if tfutils.IsRetriableErrorForEnvInstance(err) {
+				return nil, provisioning.StateUpdating, nil
+			}
+
 			if err != nil {
 				return subRes, "", err
 			}
@@ -379,6 +387,10 @@ func (rs *subaccountEnvironmentInstanceResource) Delete(ctx context.Context, req
 
 			if comRes.StatusCode == http.StatusNotFound {
 				return subRes, "DELETED", nil
+			}
+
+			if tfutils.IsRetriableErrorForEnvInstance(err) {
+				return nil, provisioning.StateDeleting, nil
 			}
 
 			if err != nil {

--- a/internal/tfutils/retry.go
+++ b/internal/tfutils/retry.go
@@ -1,0 +1,40 @@
+package tfutils
+
+import "strings"
+
+func IsRetriableErrorForEntitlement(err error) bool {
+
+	if err == nil {
+		//If no error was raised a retry is not necessary
+		return false
+	}
+
+	if strings.Contains(err.Error(), "[Error: 30004/400]") {
+		// Error code for a locking scenario - API call must be retried
+		return true
+	} else if strings.Contains(err.Error(), "[Error: 11006/429]") {
+		// Error code when hitting a rate limit - API call must be retried
+		return true
+	} else {
+		// No retry possible as error code does not indicate a valid retry scenario
+		return false
+	}
+
+}
+
+func IsRetriableErrorForEnvInstance(err error) bool {
+
+	if err == nil {
+		//If no error was raised a retry is not necessary
+		return false
+	}
+
+	if strings.Contains(err.Error(), "Command timed out. Please try again later.") {
+		// Error 504 Gateway Timeout - API call must be retried
+		return true
+	} else {
+		// No retry possible as error code does not indicate a valid retry scenario
+		return false
+	}
+
+}

--- a/internal/tfutils/retry_test.go
+++ b/internal/tfutils/retry_test.go
@@ -1,0 +1,102 @@
+package tfutils
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRetryEnvInstance(t *testing.T) {
+	type expects struct {
+		retryPossible bool
+	}
+
+	tests := []struct {
+		description string
+		uut         error
+		expects     expects
+	}{
+
+		{
+			description: "Random Error",
+			uut:         fmt.Errorf("Random Error"),
+			expects: expects{
+				retryPossible: false,
+			},
+		},
+		{
+			description: "Error 504 Gateway Timeout",
+			uut:         fmt.Errorf("Some generic errror text. Command timed out. Please try again later."),
+			expects: expects{
+				retryPossible: true,
+			},
+		},
+		{
+			description: "No Error",
+			uut:         nil,
+			expects: expects{
+				retryPossible: false,
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			result := IsRetriableErrorForEnvInstance(test.uut)
+
+			assert.Equal(t, test.expects.retryPossible, result)
+		})
+	}
+}
+
+func TestRetrySubaccountEntitlement(t *testing.T) {
+	type expects struct {
+		retryPossible bool
+	}
+
+	tests := []struct {
+		description string
+		uut         error
+		expects     expects
+	}{
+
+		{
+			description: "Random Error",
+			uut:         fmt.Errorf("Random Error"),
+			expects: expects{
+				retryPossible: false,
+			},
+		},
+		{
+			description: "Rate Limit Error",
+			uut:         fmt.Errorf("Some generic errror text. [Error: 30004/400]."),
+			expects: expects{
+				retryPossible: true,
+			},
+		},
+		{
+			description: "Locking Error",
+			uut:         fmt.Errorf("Some generic errror text. [Error: 11006/429]. Some more generic error text."),
+			expects: expects{
+				retryPossible: true,
+			},
+		},
+
+		{
+			description: "No Error",
+			uut:         nil,
+			expects: expects{
+				retryPossible: false,
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			result := IsRetriableErrorForEntitlement(test.uut)
+
+			assert.Equal(t, test.expects.retryPossible, result)
+		})
+	}
+}

